flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java

```diff
@@ -97,24 +123,26 @@ public void checkpointState(
 		// Step (3): Take the state snapshot. This should be largely asynchronous, to not
 		//           impact progress of the streaming topology
 
-		CheckpointStreamFactory storage = checkpointStorage.resolveCheckpointStorageLocation(
-			checkpointMetaData.getCheckpointId(),
-			checkpointOptions.getTargetLocation());
-
 		long startSyncPartNano = System.nanoTime();
 
 		HashMap<OperatorID, OperatorSnapshotFutures> operatorSnapshotsInProgress = new HashMap<>(operatorChain.getNumberOfOperators());
+		ChannelStateWriteResult channelStateWriteResult =
+			checkpointOptions.getCheckpointType() == CHECKPOINT ? channelStateWriter.getWriteResult(checkpointMetaData.getCheckpointId()) :
+				ChannelStateWriteResult.EMPTY;
 		try {
 			for (StreamOperatorWrapper<?, ?> operatorWrapper : operatorChain.getAllOperators(true)) {
-				StreamOperator<?> op = operatorWrapper.getStreamOperator();
-				OperatorSnapshotFutures snapshotInProgress = checkpointStreamOperator(
-					op,
-					checkpointMetaData,
-					checkpointOptions,
-					storage,
-					isCanceled);
-				operatorSnapshotsInProgress.put(op.getOperatorID(), snapshotInProgress);
+				operatorSnapshotsInProgress.put(
+					operatorWrapper.getStreamOperator().getOperatorID(),
+					buildOperatorSnapshotFutures(
+						checkpointMetaData,
+						checkpointOptions,
+						operatorChain,
+						operatorWrapper.getStreamOperator(),
+						isCanceled,
+						channelStateWriteResult)
+				);
 			}
+			checkpointStorage.clearCacheFor(checkpointMetaData.getCheckpointId());
```

Comment:
nit: I prefer to extracting `checkpointStorage.resolveCheckpointStorageLocation` from `buildOperatorSnapshotFutures` and put it before `for` loop, then it is easy to trace both the creation and clear actions in the same page.